testing/helpers/matrix-test-helper.js

```diff
@@ -1,33 +1,255 @@
+        currentNamePart = nextNamePart;
+      }
+      // eslint-disable-next-line no-continue
+      continue;
+    }
+
+    if (parseResult) {
+      parseResult.groups = parseResult.groups || {};
+      if (parseResult.groups.data && parseResult.groups.data === 'description.md') {
+        // eslint-disable-next-line no-continue
+        continue;
+      }
+
+      result.masks.push(patternGroupFromValues(
+        parseResult.groups.product,
+        parseResult.groups.demo,
+        parseResult.groups.framework,
+      ));
+    } else {
+      // eslint-disable-next-line no-console
+      if (settings.verbose) { console.log('Unable to parse changed file, running all tests: ', fileName); }
+      return undefined;
     }
   }
+
+  return result;
+}
+function getExplicitTests() {
+  const result = getExplicitTestsInternal();
+  if (result) {
+    const oldToJSON = RegExp.prototype.toJSON;
+    try {
+      // eslint-disable-next-line no-extend-native
+      RegExp.prototype.toJSON = RegExp.prototype.toString;
+      // eslint-disable-next-line no-console
+      if (settings.verbose) { console.log('Test filters: \r\n', JSON.stringify(result, null, 2)); }
+    } finally {
+      // eslint-disable-next-line no-extend-native
+      RegExp.prototype.toJSON = oldToJSON;
+    }
+  }
+
+  return result;
 }
 
 export function shouldRunFramework(currentFramework) {
-  return !currentFramework || currentFramework.toLowerCase() === targetFramework.toLowerCase();
+  return !currentFramework
+    || !settings.targetFramework
+    || currentFramework.toLowerCase() === settings.targetFramework.toLowerCase();
 }
 
 export function shouldRunTestAtIndex(testIndex) {
-  return (testIndex % total) === (current - 1);
+  return !settings.total || !settings.current || !testIndex
+    || (testIndex % settings.total) === (settings.current - 1);
 }
 
 export function shouldRunTest(currentFramework, testIndex) {
   return shouldRunFramework(currentFramework) && shouldRunTestAtIndex(testIndex);
 }
 
+export function shouldRunTestExplicitly(demoUrl) {
+  if (!settings.explicitTests) { return false; }
+
+  const parts = demoUrl.split('/').filter((x) => x && x.length);
+
+  return shouldRunTestExplicitlyInternal(
+    parts[parts.length - 1],
+    parts[parts.length - 3],
+    parts[parts.length - 2],
+  );
+}
+
+export function runTestAt(test, demoUrl) {
+  let executor = test;
+  if (settings.explicitTests) {
+    if (shouldRunTestExplicitly(demoUrl)) { executor = test.only; } else { executor = test.skip; }
+  }
+  return executor.page(demoUrl);
+}
+
+export function runTest(testObject, framework, product, demo, callback) {
+  const index = settings.manualTestIndex;
+  settings.manualTestIndex += 1;
+  if (settings.explicitTests) {
+    if (shouldRunTestExplicitlyInternal(framework, product, demo)) {
+      callback(testObject.only);
+      return;
+    }
+    if (settings.explicitTests.traceTree) {
+      const stackInterestPoints = getCallStackInterestPoints();
```

Comment:
May be it is possible to get test file path from taskcafe instead of using call stack magic?
Test args have interesting undocumented parameters: https://github.com/DevExpress/testcafe/issues/2826#issuecomment-520950150. Also we can ask TestCafe team about it.

Reply:
We have to access to a test instance object at the current point of execution, we have only access to the test factory object. Test factory does not store any information about location etc.